File: apps/backend/app/modules/recruitment/schemas/tasks.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from app.common.utils.datetime_utils import normalize_datetime
from app.modules.recruitment.enums.activity_type import ActivityType
from app.modules.recruitment.models import TaskAssignmentType

DATE_ERROR_MSG = "due_date must be greater than or equal to start_date"


class TaskCreate(BaseModel):
    title: str
    description: str | None = None
    tracked_activity_type: ActivityType
    target_count: int = Field(gt=0)
    assignee_type: TaskAssignmentType
    assignee_id: str | None = None
    start_date: datetime
    due_date: datetime

    @model_validator(mode="after")
    def check_dates(self) -> "TaskCreate":
        if self.start_date and self.due_date:
            s_dt = normalize_datetime(self.start_date)
            d_dt = normalize_datetime(self.due_date)
            if d_dt < s_dt:
                raise ValueError(DATE_ERROR_MSG)
        return self


class TaskUpdate(BaseModel):
    is_active: bool


class RecruiterProgress(BaseModel):
    employee_id: str
    name: str
    completed_count: int
    progress_percentage: int


class TaskResponse(BaseModel):
    id: str
    title: str
    description: str | None
    tracked_activity_type: ActivityType
    target_count: int
    assignee_type: TaskAssignmentType
    assignee_id: str | None
    start_date: datetime
    due_date: datetime

    @model_validator(mode="after")
    def check_dates(self) -> "TaskResponse":
        if self.start_date and self.due_date:
            s_dt = normalize_datetime(self.start_date)
            d_dt = normalize_datetime(self.due_date)
            if d_dt < s_dt:
                raise ValueError(DATE_ERROR_MSG)
        return self

    is_active: bool
    created_at: datetime

    # Progress for the viewing user (if single) or overall summary
    completed_count: int = 0
    progress_percentage: int = 0

    # Detailed progress if queried by Admin for a Team/All task
    detailed_progress: list[RecruiterProgress] | None = None


class TaskUpdatePayload(BaseModel):
    title: str | None = None
    description: str | None = None
    tracked_activity_type: ActivityType | None = None
    target_count: int | None = Field(None, gt=0)
    assignee_type: TaskAssignmentType | None = None
    assignee_id: str | None = None
    start_date: datetime | None = None
    due_date: datetime | None = None

    @model_validator(mode="after")
    def check_dates(self) -> "TaskUpdatePayload":
        if self.start_date and self.due_date:
            s_dt = normalize_datetime(self.start_date)
            d_dt = normalize_datetime(self.due_date)
            if d_dt < s_dt:
                raise ValueError(DATE_ERROR_MSG)
        return self

File: apps/frontend/src/lib/api/tasks.ts
type ApiFetch = <T>(path: string, options?: RequestInit) => Promise<T>;

export type TaskAssignmentType = "single" | "team" | "all";
export type TrackedActivityType =
  | "mapped"
  | "stage_moved"
  | "offer_sent"
  | "offer_accepted"
  | "joined"
  | "rejected"
  | "unmapped";

export interface RecruiterProgress {
  employee_id: string;
  name: string;
  completed_count: number;
  progress_percentage: number;
}

export interface TaskResponse {
  id: string;
  title: string;
  description: string | null;
  tracked_activity_type: TrackedActivityType;
  target_count: number;
  assignee_type: TaskAssignmentType;
  assignee_id: string | null;
  start_date: string;
  due_date: string;
  is_active: boolean;
  created_at: string;
  completed_count: number;
  progress_percentage: number;
  detailed_progress: RecruiterProgress[] | null;
}

export interface TaskCreatePayload {
  title: string;
  description?: string;
  tracked_activity_type: TrackedActivityType;
  target_count: number;
  assignee_type: TaskAssignmentType;
  assignee_id?: string;
  start_date: string;
  due_date: string;
}

export async function listTasks(apiFetch: ApiFetch): Promise<TaskResponse[]> {
  return await apiFetch<TaskResponse[]>("/api/v1/tasks/");
}

export async function createTask(
  apiFetch: ApiFetch,
  payload: TaskCreatePayload,
): Promise<TaskResponse> {
  return await apiFetch<TaskResponse>("/api/v1/tasks/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function deleteTask(apiFetch: ApiFetch, taskId: string): Promise<void> {
  await apiFetch<void>(`/api/v1/tasks/${taskId}`, { method: "DELETE" });
}
